File: rust/src/models/credit_note_event_data.rs
// this file is @generated
use serde::{Deserialize, Serialize};

use super::{
    credit_note_id::CreditNoteId, credit_note_status::CreditNoteStatus, customer_id::CustomerId,
    invoice_id::InvoiceId, invoice_line_item::InvoiceLineItem,
    tax_breakdown_item::TaxBreakdownItem,
};

#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize)]
pub struct CreditNoteEventData {
    pub created_at: String,

    pub credit_note_id: CreditNoteId,

    /// Absent while the credit note is a draft — the number is assigned at finalization.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credit_note_number: Option<String>,

    pub credited_amount_cents: i32,

    pub currency: String,

    /// User-defined custom property values, keyed by definition key.
    pub custom_properties: serde_json::Value,

    pub customer_id: CustomerId,

    pub invoice_id: InvoiceId,

    /// Number of the invoice being credited.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub invoice_number: Option<String>,

    /// Credited line items (negated amounts).
    pub line_items: Vec<InvoiceLineItem>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub memo: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,

    pub refunded_amount_cents: i32,

    pub status: CreditNoteStatus,

    pub subtotal: i32,

    pub tax_amount: i32,

    /// Per-rate tax (VAT) breakdown for the credited amount.
    pub tax_breakdown: Vec<TaxBreakdownItem>,

    pub total: i32,
}

impl CreditNoteEventData {
    pub fn new(
        created_at: String,
        credit_note_id: CreditNoteId,
        credited_amount_cents: i32,
        currency: String,
        custom_properties: serde_json::Value,
        customer_id: CustomerId,
        invoice_id: InvoiceId,
        line_items: Vec<InvoiceLineItem>,
        refunded_amount_cents: i32,
        status: CreditNoteStatus,
        subtotal: i32,
        tax_amount: i32,
        tax_breakdown: Vec<TaxBreakdownItem>,
        total: i32,
    ) -> Self {
        Self {
            created_at,
            credit_note_id,
            credit_note_number: None,
            credited_amount_cents,
            currency,
            custom_properties,
            customer_id,
            invoice_id,
            invoice_number: None,
            line_items,
            memo: None,
            reason: None,
            refunded_amount_cents,
            status,
            subtotal,
            tax_amount,
            tax_breakdown,
            total,
        }
    }
}

File: typescript/src/models/creditNoteEventData.ts
// this file is @generated
import { parseDateTime } from "../datetime";
import { type CreditNoteId, CreditNoteIdSerializer } from "./creditNoteId";
import { type CreditNoteStatus, CreditNoteStatusSerializer } from "./creditNoteStatus";
import { type CustomerId, CustomerIdSerializer } from "./customerId";
import { type InvoiceId, InvoiceIdSerializer } from "./invoiceId";
import { type InvoiceLineItem, InvoiceLineItemSerializer } from "./invoiceLineItem";
import { type TaxBreakdownItem, TaxBreakdownItemSerializer } from "./taxBreakdownItem";

export interface CreditNoteEventData {
  createdAt: Date;

  creditNoteId: CreditNoteId;

  /** Absent while the credit note is a draft — the number is assigned at finalization. */
  creditNoteNumber?: string | null;

  creditedAmountCents: number;

  currency: string;

  /** User-defined custom property values, keyed by definition key. */
  customProperties: unknown;

  customerId: CustomerId;

  invoiceId: InvoiceId;

  /** Number of the invoice being credited. */
  invoiceNumber?: string | null;

  /** Credited line items (negated amounts). */
  lineItems: InvoiceLineItem[];

  memo?: string | null;

  reason?: string | null;

  refundedAmountCents: number;

  status: CreditNoteStatus;

  subtotal: number;

  taxAmount: number;

  /** Per-rate tax (VAT) breakdown for the credited amount. */
  taxBreakdown: TaxBreakdownItem[];

  total: number;
}

export const CreditNoteEventDataSerializer = {
  _fromJsonObject(object: any): CreditNoteEventData {
    return {
      createdAt: parseDateTime(object["created_at"]),
      creditNoteId: CreditNoteIdSerializer._fromJsonObject(object["credit_note_id"]),
      creditNoteNumber: object["credit_note_number"],
      creditedAmountCents: object["credited_amount_cents"],
      currency: object["currency"],
      customProperties: object["custom_properties"],
      customerId: CustomerIdSerializer._fromJsonObject(object["customer_id"]),
      invoiceId: InvoiceIdSerializer._fromJsonObject(object["invoice_id"]),
      invoiceNumber: object["invoice_number"],
      lineItems: object["line_items"].map((item: any) =>
        InvoiceLineItemSerializer._fromJsonObject(item)
      ),
      memo: object["memo"],
      reason: object["reason"],
      refundedAmountCents: object["refunded_amount_cents"],
      status: CreditNoteStatusSerializer._fromJsonObject(object["status"]),
      subtotal: object["subtotal"],
      taxAmount: object["tax_amount"],
      taxBreakdown: object["tax_breakdown"].map((item: any) =>
        TaxBreakdownItemSerializer._fromJsonObject(item)
      ),
      total: object["total"],
    };
  },

  _toJsonObject(self: CreditNoteEventData): any {
    return {
      created_at: self.createdAt,
      credit_note_id: CreditNoteIdSerializer._toJsonObject(self.creditNoteId),
      credit_note_number: self.creditNoteNumber,
      credited_amount_cents: self.creditedAmountCents,
      currency: self.currency,
      custom_properties: self.customProperties,
      customer_id: CustomerIdSerializer._toJsonObject(self.customerId),
      invoice_id: InvoiceIdSerializer._toJsonObject(self.invoiceId),
      invoice_number: self.invoiceNumber,
      line_items: self.lineItems.map((item: any) =>
        InvoiceLineItemSerializer._toJsonObject(item)
      ),
      memo: self.memo,
      reason: self.reason,
      refunded_amount_cents: self.refundedAmountCents,
      status: CreditNoteStatusSerializer._toJsonObject(self.status),
      subtotal: self.subtotal,
      tax_amount: self.taxAmount,
      tax_breakdown: self.taxBreakdown.map((item: any) =>
        TaxBreakdownItemSerializer._toJsonObject(item)
      ),
      total: self.total,
    };
  },
};
